### PR TITLE
Developer can declare steps in a test, reflected in the report

### PR DESCRIPTION
**Pull request by vanhoofmaarten**
The user can declare steps in the tests. These step-declarations are visible in the report of the test.

### CODE REVIEW

File: src/testcafe/step.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    // @ts-ignore
+    let { meta } = testController.testRun.test;
+    if (!meta) {
+      meta = {};
+      testController.testRun.test.meta = meta;
+    }
+    return meta;
+  }
+
+  // // Returns a new copy of the reporter thus may not be able to be used.
```
Comment (vanhoofmaarten) [inline src/testcafe/step.ts:45]: Will not be used, please remove.

File: src/testcafe/step.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  // private getReporter(testController: TestController): AllureReporter {
+  //   // @ts-ignore
+  //   return testController.testRun.opts.reporter[0].name().getAllureReporter();
+  // }
+}
+
+/* The TestController loses its parameters when returned as a TestControllerPromise. 
+   Therefore the steps cannot be added without a clean TestController.
+*/
+export default async function step(name: string, testController: TestController, stepAction: any) {
```
Comment (vanhoofmaarten) [inline src/testcafe/step.ts:55]: Please import step into `index.ts` and export it as a named function from there.

File: src/utils/config.ts [hunk 3]
```diff
@@ -26,6 +27,7 @@ const defaultReporterConfig = {
     FLAKY: 'Flaky test',
     SCREENSHOT_MANUAL: 'Screenshot taken manually',
     SCREENSHOT_ON_FAIL: 'Screenshot taken on fail',
+    DEFUALT_STEP_NAME: 'Test step',
```
Comment (vanhoofmaarten) [inline src/utils/config.ts:30]: Typo

File: tests/e2e/example.spec.ts [hunk 4]
```diff
@@ -1,88 +1,109 @@
...
+})('Example severity metadata override MINOR', async (t) => {
   await t
     .typeText('#developer-name', 'John Smith')
     .click('#submit-button')
     .expect(Selector('#article-header').innerText)
-    .eql('Thank you, Jane Smith!');
+    .eql('Thank you, John Smith!');
 });
 
-test('My actual flaky e2e test', async (t) => {
```
Comment (vanhoofmaarten) [inline tests/e2e/example.spec.ts:79]: We need more certainty that the flaky test is flaky. (Read remove false positives).
Can we make an assertion on a persisted variable that is updated the first 2 attempts and that will pass the assertion the third time?

File: src/testcafe/models.ts [hunk 5]
```diff
@@ -1,3 +1,17 @@
+export {};
+declare global {
+  interface TestController {
```
Comment (vanhoofmaarten) [inline src/testcafe/models.ts:3]: Please provide link to GH issues why we need to provide our own interfaces for certain TestCafé types/interfaces and why we're not declaring the entire interface.

Comment (ronaldvandenbroek) [reply src/testcafe/models.ts:4]: Added the GH issue detailing why the TestController interface is needed and why dexlaring the entire interface is unnessesary.

File: tests/e2e/example.spec.ts (hunk 4 above)
Comment (ronaldvandenbroek) [reply tests/e2e/example.spec.ts:79]: Reworked the not-consistently-flaky Flaky test to be more consistent in it's flakiness. (By adding a persisted variable that is increased by one each run and only matches on the third run)

File: src/utils/config.ts (hunk 3 above)
Comment (ronaldvandenbroek) [reply src/utils/config.ts]: Renamed to `DEFAULT_STEP_NAME`.

File: src/testcafe/step.ts (hunk 1 above)
Comment (ronaldvandenbroek) [reply src/testcafe/step.ts:44]: Removed the unused code.

File: src/reporter/allure-reporter.ts [hunk 6]
```diff
@@ -141,28 +142,73 @@ export default class AllureReporter {
...
+      this.addScreenshotAttachments(currentTest, testRunInfo);
+    }
+
     currentTest.detailsMessage = testMessages;
     currentTest.detailsTrace = testDetails;
     currentTest.stage = Stage.FINISHED;
     currentTest.endTest();
   }
 
+  private addStepsWithAttachments(test: AllureTest, testRunInfo: TestRunInfo, steps: TestStep[]) {
```
[the remark this answers is not in the document]
Comment (ronaldvandenbroek) [reply src/reporter/allure-reporter.ts:164]: Added comment to clarify the functionality.

File: src/utils/config.ts [hunk 7]
```diff
@@ -26,6 +27,7 @@ const defaultReporterConfig = {
     FLAKY: 'Flaky test',
     SCREENSHOT_MANUAL: 'Screenshot taken manually',
     SCREENSHOT_ON_FAIL: 'Screenshot taken on fail',
+    DEFAULT_STEP_NAME: 'Test step',
```
Comment (vanhoofmaarten) [inline src/utils/config.ts:30]: I think just `step` will suffice

**Pull request merged**